prow/plank/reconciler.go

```diff
@@ -781,29 +792,39 @@ func (r *reconciler) canExecuteConcurrently(ctx context.Context, pj *prowv1.Prow
 	}
 	r.log.Infof("got %d not completed with same name", len(pjs.Items))
 
-	var pendingOrOlderMatchingPJs int
-	for _, foundPJ := range pjs.Items {
-		// Ignore self here.
-		if foundPJ.UID == pj.UID {
-			continue
-		}
-		if foundPJ.Status.State == prowv1.PendingState {
-			pendingOrOlderMatchingPJs++
-			continue
-		}
+	pendingOrOlderMatchingPJs := countPendingOrOlderMatchingPJs(*pj, pjs.Items)
+	if pendingOrOlderMatchingPJs >= pj.Spec.MaxConcurrency {
+		r.log.WithFields(pjutil.ProwJobFields(pj)).
+			Debugf("Not starting another instance of %s, have %d instances that are pending or older, %d is the limit",
+				pj.Spec.Job, pendingOrOlderMatchingPJs, pj.Spec.MaxConcurrency)
+		return false, nil
+	}
 
-		// At this point we know that foundPJ is in Triggered state, if its older than our prowJobs it gets
-		// priorized to make sure we execute jobs in creation order.
-		if foundPJ.CreationTimestamp.Before(&pj.CreationTimestamp) {
-			pendingOrOlderMatchingPJs++
-		}
+	return true, nil
+}
 
+func (r *reconciler) canExecuteConcurrentlyPerQueue(ctx context.Context, pj *prowv1.ProwJob) (bool, error) {
+	queueName := pj.Spec.JobQueueName
+	if queueName == "" {
+		return true, nil
 	}
 
-	if pendingOrOlderMatchingPJs >= pj.Spec.MaxConcurrency {
+	queueConcurrency, queueDefined := r.config().Plank.JobQueueConcurrencies[queueName]
+	if !queueDefined || queueConcurrency == 0 {
+		return true, nil
+	}
+
+	pjs := &prowv1.ProwJobList{}
+	if err := r.pjClient.List(ctx, pjs, optPendingProwJobs(), optJobsInQueue(queueName)); err != nil {
```

Comment:
`optPendingProwJobs()` seems overly restrictive. That will omit triggered ProwJobs and cause us to undercount in `countPendingOrOlderMatchingPJs` (there will be no older triggered PJs).

Reply:
Fixed in https://github.com/kubernetes/test-infra/pull/25674/commits/55d9a8ef93e1eccd103b77a00f96d7a84cf0cbbd. Added a `pendingTriggeredIndexKeyByJobQueueName` selector.